File: src/main/java/com/finalproject/walktogetherapi/mapping/CollectionMapping.java
package com.finalproject.walktogetherapi.mapping;

import com.finalproject.walktogetherapi.entities.*;
import com.finalproject.walktogetherapi.service.CollectionService;
import com.finalproject.walktogetherapi.service.RewardService;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Random;

public class CollectionMapping {

    private static CollectionMapping instance;

    public static CollectionMapping getInstance() {
        if (instance == null) {
            instance = new CollectionMapping();
        }
        return instance;
    }

    public List<HashMap<String, Object>> getRewardList(List<Collection> collectionList) {
        List<HashMap<String, Object>> map = new ArrayList<>();
        for (Collection collection : collectionList) {
            HashMap<String, Object> objectHashMap = new HashMap<>();
            objectHashMap.put("reward", collection.getReward());
            objectHashMap.put("isReceive", collection.isReceive());
            objectHashMap.put("isLock", collection.isLock());
            map.add(objectHashMap);
        }
        return map;
    }

    public Reward randomReward(int level, List<Reward> rewardList) {
        List<Reward> result = new ArrayList<>();
        for (Reward reward : rewardList) {
            if (reward.getLevel() <= level)
                result.add(reward);
        }
        return result.get(new Random().nextInt(result.size()));
    }

    public void addCollection(RewardService rewardService, CollectionService collectionService, Patient patient) {
        List<Reward> rewardList = rewardService.findAll();
        for (Reward reward : rewardList) {
                Collection collection = new Collection();
                collection.setReceive(false);
                collection.setLock(true);
                collection.setPatient(patient);
                collection.setReward(reward);
                collectionService.create(collection);
            }
    }

    public void unlockReward(CollectionService collectionService, Patient patient) {
        List<Collection> collectionList = collectionService.findByPatientId(patient.getId());
        for (Collection collection : collectionList) {
            collection.setLock(false);
            collectionService.update(collection.getId(), collection);
        }
    }
}

File: src/main/java/com/finalproject/walktogetherapi/mapping/CollectionMapping.kt
package com.finalproject.walktogetherapi.mapping

import com.finalproject.walktogetherapi.entities.*
import com.finalproject.walktogetherapi.entities.Collection
import com.finalproject.walktogetherapi.service.CollectionService
import com.finalproject.walktogetherapi.service.RewardService

import java.util.ArrayList
import java.util.HashMap
import java.util.Random

class CollectionMapping {

    fun getAlbum(patient: Patient, collectionService: CollectionService, range: Int): List<HashMap<String, Any>> {
        val map = ArrayList<HashMap<String, Any>>()

        var i = 1
        while (i <= 30 - range) {
            val j = i + range
            val objectHashMap = HashMap<String, Any>()
            val rewardList = collectionService.findByRangeLevel(patient.id, i, j).sortedWith(compareBy({ it.reward.level }))
            objectHashMap["albumName"] = "เลเวล $i-$j"
            objectHashMap["collection"] = rewardList
            objectHashMap["isLock"] = !(patient.level > j || patient.level in i..j)
            objectHashMap["previewImage"] = rewardList.filter {
                it.reward.level == i || it.reward.level == j
            }.take(4)
            map.add(objectHashMap)
            i += (range + 1)
        }
        return map
    }

    fun getRewardList(collectionList: List<Collection>): List<HashMap<String, Any>> {
        val map = ArrayList<HashMap<String, Any>>()
        for (collection in collectionList) {
            val objectHashMap = HashMap<String, Any>()
            objectHashMap["reward"] = collection.reward
            objectHashMap["isReceive"] = collection.isReceive
            objectHashMap["isLock"] = collection.isLock
            map.add(objectHashMap)
        }
        return map
    }


    fun randomReward(level: Int, rewardList: List<Reward>): Reward {
        val result = rewardList.filterTo(ArrayList()) { it.level <= level }
        return result[Random().nextInt(result.size)]
    }

    fun addCollection(rewardService: RewardService, collectionService: CollectionService, patient: Patient) {
        val rewardList = rewardService.findAll()
        for (reward in rewardList) {
            val collection = Collection()
            collection.isReceive = false
            collection.isLock = reward.level != 1
            collection.patient = patient
            collection.reward = reward
            collectionService.create(collection)
        }
    }

    fun receiveReward(collectionService: CollectionService, patient: Patient, reward: Reward) {
        val collection = collectionService.findByRewardId(reward.id, patient.id)
        collection.isReceive = true
        collection.isLock = false
        collectionService.update(collection.id, collection)
    }

    fun unlockReward(collectionService: CollectionService, patient: Patient) {
        val collectionList = collectionService.findByPatientId(patient.id)
        for (collection in collectionList) {
            if (patient.level >= collection.reward.level) {
                collection.isLock = false
                collectionService.update(collection.id, collection)
            }
        }
    }

    companion object {

        private var instance: CollectionMapping? = null

        fun getInstance(): CollectionMapping {
            if (instance == null) {
                instance = CollectionMapping()
            }
            return instance as CollectionMapping
        }
    }
}
